File: B. Medium/0840-magic-squares-in-grid/0840-magic-squares-in-grid.c
bool checkGridMagic(int grid[3][3]) {
    // Check if all numbers from 1 to 9 are present
    bool seen[10] = {false}; // Index 0 is unused
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            int num = grid[i][j];
            if (num < 1 || num > 9 || seen[num]) {
                return false;
            }
            seen[num] = true;
        }
    }

    // Check if all row, column, and diagonal sums are the same
    int sum = grid[0][0] + grid[0][1] + grid[0][2]; // Reference sum

    // Check rows
    for (int i = 1; i < 3; i++) {
        int rowSum = grid[i][0] + grid[i][1] + grid[i][2];
        if (rowSum != sum) return false;
    }

    // Check columns
    for (int j = 0; j < 3; j++) {
        int colSum = grid[0][j] + grid[1][j] + grid[2][j];
        if (colSum != sum) return false;
    }

    // Check diagonals
    int diag1Sum = grid[0][0] + grid[1][1] + grid[2][2];
    int diag2Sum = grid[0][2] + grid[1][1] + grid[2][0];
    if (diag1Sum != sum || diag2Sum != sum) return false;

    return true;
}

int numMagicSquaresInside(int** grid, int gridSize, int* gridColSize) {
    int gridCount = 0;

    // Iterate over all possible 3x3 sub-grids
    for (int i = 0; i <= gridSize - 3; i++) {
        for (int j = 0; j <= gridColSize[i] - 3; j++) {
            int subGrid[3][3];
            for (int x = 0; x < 3; x++) {
                for (int y = 0; y < 3; y++) {
                    subGrid[x][y] = grid[i + x][j + y];
                }
            }
            if (checkGridMagic(subGrid)) {
                gridCount += 1;
            }
        }
    }

    return gridCount;
}

File: B. Medium/0840-magic-squares-in-grid/0840-magic-squares-in-grid.py
class Solution:
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        def checkGridMagic(grid: List[List[int]]) -> bool:
            flattenedGridSet = set([num for row in grid for num in row])
            if flattenedGridSet != set([1, 2, 3, 4, 5, 6, 7, 8, 9]):
                return False
            gridRowSum = [sum(row) for row in grid]
            gridColSum = [sum(col) for col in zip(*grid)]
            gridDiagSum = list()
            gridDiagSum.append(sum(grid[i][i] for i in range(3)))
            gridDiagSum.append(sum(grid[i][2 - i] for i in range(3)))
            gridRowColDiagSum = gridRowSum + gridColSum + gridDiagSum
            return all(x == gridRowColDiagSum[0] for x in gridRowColDiagSum)

        gridCount = 0
        checkSum = sum(range(10))
        for i in range(len(grid)-2):
            for j in range(len(grid[i])-2):
                subGrid = [row[j:j+3] for row in grid[i:i+3]]
                if checkGridMagic(subGrid):
                    gridCount += 1
        return gridCount

